File: csharp/DVSClient/DVSClient.Tests/layout/AddressLayoutTests.cs
using DVSClient.Address.Format.Enrichment;
using DVSClient.Address.Layout.Attributes;
using DVSClient.Address.Layout.Elements;
using DVSClient.Exceptions;
using DVSClientTests;
using NUnit.Framework;

namespace DVSClient.Address.Layout.Tests
{
    /* 
    * These tests are dependent on pre-existing layouts being available.
    */
    [TestFixture]
    public class AddressLayoutTests
    {
        [OneTimeSetUp]
        public void TestSetup()
        {
            // Some of these tests rely on pre-existing layouts because creating a layout during the tests is not feasible
            // This tests that they exist
            var testLayout = GetLayout(Setup.ExistingTestLayout);

            if (testLayout?.Error?.Title == "Not Found")
            {
                // Doesn't exist yet. Create it.
                CreateTestLayout();
                Assert.Fail($"The layout {Setup.ExistingTestLayout} did not exist. This has now been created but you will need to wait for the creation to complete (can be 10 minutes or so)");
            }
            else if (testLayout?.Layout?.Status != Status.Completed)
            {
                Assert.Fail($"The layout {Setup.ExistingTestLayout} is not complete. Please wait for it to complete before running these tests (can be 10 minutes or so)");
            }

            // Clean up any test layout from previous runs (they need to have completed creation before they can be deleted
            DeleteTestLayouts();
        }

        [OneTimeTearDown]
        public void TestCleanup()
        {
            // Delete any layout created during the tests with the common prefix
            DeleteTestLayouts();
        }

        [Test]
        public void Layout_Get()
        {
            var configuration = Configuration.NewBuilder(Setup.ValidTokenAddress).Build();
            var client = ExperianDataValidation.GetAddressLayoutClient(configuration);
            var result = client.GetLayout(Setup.ExistingTestLayout);
            Assert.That(result.Error, Is.Null);
        }


        [Test]
        public void Layout_Create()
        {
            var configuration = Configuration.NewBuilder(Setup.ValidTokenAddress).Build();
            var client = ExperianDataValidation.GetAddressLayoutClient(configuration);
            var line1 = new LayoutLineVariable("addr_line_1");
            var line2 = new LayoutLineVariable("addr_line_2");
            var line3 = new LayoutLineFixed("post_code", Aus.PostalCode);
            var line4 = new LayoutLineFixed("country_name", Aus.CountryName);
            var layoutName = GetUniqueLayoutName();
            var createLayoutResult = client.CreateLayout(layoutName, new List<LayoutLineVariable> { line1, line2 }, new List<LayoutLineFixed> { line3, line4 }, Dataset.AuAddress);
            Assert.That(createLayoutResult.Error, Is.Null);
            var result = client.GetLayout(layoutName);
            Assert.That(result.Error, Is.Null);
        }

        [Test]
        public void Layout_Create_WithOptions()
        {
            var configuration = Configuration.NewBuilder(Setup.ValidTokenAddress)
                .SetTransactionId(Guid.NewGuid().ToString())
                .Build();
            var client = ExperianDataValidation.GetAddressLayoutClient(configuration);
            // TODO
            // Add your test logic here
        }

        [Test]
        public void Layout_Delete_DoesNotExist_Throws()
        {
            var configuration = Configuration.NewBuilder(Setup.ValidTokenAddress).Build();
            var client = ExperianDataValidation.GetAddressLayoutClient(configuration);
            var layoutName = "ThisLayoutDoesntExist";

            var ex = Assert.Throws<NotFoundException>(() => client.DeleteLayout(layoutName));
        }

        [Test]
        public void Format_WithCustomLayout_WithComponents()
        {
            var configuration = Address.Configuration
                .NewBuilder(Setup.ValidTokenAddress)
                .UseDataset(Dataset.GbAddress)
                .UseLayout(Setup.ExistingTestLayout)
                .IncludeComponents()
                .Build();
            var client = ExperianDataValidation.GetAddressClient((Address.Configuration)configuration);
            var searchResultAutoComplete = client.Search(SearchType.Autocomplete, "56 Queens R");
            var formatResult = client.Format(searchResultAutoComplete.Suggestions.First().GlobalAddressKey);
            Assert.That(formatResult.Confidence, Is.EqualTo(Confidence.VerifiedMatch));
            //Assert.That(formatResult.GlobalAddressKey, Is.EqualTo(searchResultAutoComplete.Suggestions.First().GlobalAddressKey));
            Assert.That(formatResult.AddressFormatted, Is.Not.Null);
            Assert.That(formatResult.AddressFormatted?.LayoutName, Is.EqualTo(Setup.ExistingTestLayout));
            Assert.That(formatResult.AddressFormatted?.HasEnoughLines, Is.Null);
            Assert.That(formatResult.AddressFormatted?.HasTruncatedLines, Is.Null);
            Assert.That(formatResult.AddressFormatted?.HasMissingSubPremises, Is.Null);
            Assert.That(formatResult.AddressFormatted?.Address.Count, Is.EqualTo(4));
            Assert.That(formatResult.AddressFormatted?.Address["addr_line_1"], Is.Not.Empty);
            Assert.That(formatResult.AddressFormatted?.Address["addr_line_2"], Is.Not.Empty);
            Assert.That(formatResult.AddressFormatted?.Address["post_code"], Is.Not.Empty);
            Assert.That(formatResult.AddressFormatted?.Address["country_name"], Is.Not.Empty);

            Assert.That(formatResult.Components, Is.Not.Null);
            Assert.That(formatResult.Components?.PostalCode?.FullName, Is.EqualTo(formatResult.AddressFormatted?.Address["post_code"].ToUpperInvariant()));
        }

        [Test]
        public void Suggestions_Format_WithCustomLayout()
        {
            var configuration = Address.Configuration
                .NewBuilder(Setup.ValidTokenAddress)
                .SetTransactionId(Guid.NewGuid().ToString())
                .UseDataset(Dataset.AuAddress)
                .UseLayout(Setup.ExistingTestLayout)
                .Build();
            var client = ExperianDataValidation.GetAddressClient(configuration);

            var searchResult = client.SuggestionsFormat("onslow st, perth");

            Assert.That(searchResult.Suggestions, Is.Not.Empty);

            // When using a custom layout the Address objects should all be empty
            Assert.That(searchResult.Suggestions.All(x => x.Address == null));
            // When using a custom layout the AddressFormatted objects should all be empty
            Assert.That(searchResult.Suggestions.All(x => x.AddressFormatted != null));
        }

        private string GetUniqueLayoutName()
        {
            return Setup.TestLayoutPrefix + Guid.NewGuid().ToString();
        }

        private GetLayoutResult GetLayout(string layoutName)
        {
            var configuration = Configuration.NewBuilder(Setup.ValidTokenAddress).Build();
            var client = ExperianDataValidation.GetAddressLayoutClient(configuration);
            return client.GetLayout(layoutName);
        }

        private void CreateTestLayout()
        {
            var configuration = Configuration.NewBuilder(Setup.ValidTokenAddress).Build();
            var client = ExperianDataValidation.GetAddressLayoutClient(configuration);
            var line1 = new LayoutLineVariable("addr_line_1");
            var line2 = new LayoutLineVariable("addr_line_2");
            var line3 = new LayoutLineFixed("post_code", new List<IAddressElement?> { Aus.PostalCode, Gbr.PostalCode.Postcode });
            var line4 = new LayoutLineFixed("country_name", new List<IAddressElement?> { Aus.CountryName, Gbr.Country.Country });
            var layoutName = Setup.ExistingTestLayout;
            var createLayoutResult = client.CreateLayout(
                layoutName, 
                new List<LayoutLineVariable> { line1, line2 },
                new List<LayoutLineFixed> { line3, line4 },
                Dataset.AuAddress, Dataset.GbAddress);
            Assert.That(createLayoutResult.Error, Is.Null);
        }

        private void DeleteTestLayouts()
        {
            // Get all layouts starting with the predefined prefix
            var configuration = Configuration.NewBuilder(Setup.ValidTokenAddress).Build();
            var client = ExperianDataValidation.GetAddressLayoutClient(configuration);
            GetLayoutListResult layoutsResult;

            try
            {
                layoutsResult = client.GetLayouts(null, new List<Dataset>(), Setup.TestLayoutPrefix);
            }
            catch (NotFoundException)
            {
                // No layouts to delete
                return;
            }

            // Delete them
            foreach (var layout in layoutsResult.Layouts)
            {
                if (layout.Status == Status.Completed && layout.Name != Setup.ExistingTestLayout)
                {
                    try
                    {
                        client.DeleteLayout(layout.Name);
                    }
                    catch (EDVSException)
                    {
                        Console.WriteLine($"Failed to delete layout {layout.Name}");
                    }
                }
            }
        }
    }
}

File: java/src/test/java/com/experian/dvs/client/layout/AddressLayoutTests.java
package com.experian.dvs.client.layout;

import com.experian.dvs.client.ExperianDataValidation;
import com.experian.dvs.client.Setup;
import com.experian.dvs.client.address.*;
import com.experian.dvs.client.address.format.Result;
import com.experian.dvs.client.address.layout.*;
import com.experian.dvs.client.address.layout.elements.Aus;
import com.experian.dvs.client.address.layout.elements.Gbr;
import com.experian.dvs.client.exceptions.EDVSException;
import com.experian.dvs.client.exceptions.NotFoundException;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.fail;

/**
 * These tests are dependent on pre-existing layouts being available.
 */
public class AddressLayoutTests {

    @BeforeAll
    static void testSetup() {
        //Some of these tests rely on pre-existing layouts because creating a layout during the tests is not feasible
        //This tests that they exist
        final GetLayoutResult layout1 = getLayout(Setup.EXISTING_TEST_LAYOUT);
        if (layout1.getLayout() == null) {
            //Doesn't exist yet. Create it.
            createLayout1();
            fail("The layout " + Setup.EXISTING_TEST_LAYOUT + " did not exist. This has now been created but you will need to wait for the creation to complete (can be 10 minutes or so)");
        }
        if (layout1.getLayout().getStatus() != Status.COMPLETED) {
            fail("The layout " + Setup.EXISTING_TEST_LAYOUT + " is not complete. Please wait for it to complete before running these tests (can be 10 minutes or so)");
        }

        //Clean up any test layout from previous runs (they need to have completed creation before they can be deleted
        deleteTestLayouts();

    }

    @AfterAll
    static void testCleanup() {
        //Delete any layout created during the tests with the common prefix
        deleteTestLayouts();
    }

    @Test
    void layout_Get() {

        final AddressLayoutConfiguration configuration = AddressLayoutConfiguration.newBuilder(Setup.VALID_TOKEN_ADDRESS).build();

        final AddressLayoutClient client = ExperianDataValidation.getAddressLayoutClient(configuration);
        final GetLayoutResult result = client.getLayout(Setup.EXISTING_TEST_LAYOUT);
        assertThat(result.getError()).isNotPresent();
    }

    @Test
    void layout_Create() {

        final AddressLayoutConfiguration configuration = AddressLayoutConfiguration
                .newBuilder(Setup.VALID_TOKEN_ADDRESS)
                .build();
        final AddressLayoutClient client = ExperianDataValidation.getAddressLayoutClient(configuration);

        final LayoutLineVariable line1 = new LayoutLineVariable("addr_line_1");
        final LayoutLineVariable line2 = new LayoutLineVariable("addr_line_2");
        final LayoutLineFixed line3 = new LayoutLineFixed("post_code", Aus.POSTAL_CODE);
        final LayoutLineFixed line4 = new LayoutLineFixed("country_name", Aus.COUNTRY_NAME);

        final String layoutName = getUniqueLayoutName();
        final CreateLayoutResult createLayoutResult = client.createLayout(layoutName,
                List.of(line1, line2),
                List.of(line3, line4),
                Dataset.AU_ADDRESS);

        assertThat(createLayoutResult.getError()).isNotPresent();
        //Check if the layout was created
        final GetLayoutResult result = client.getLayout(layoutName);
        assertThat(result.getLayout().getStatus()).isEqualTo(Status.CREATION_IN_PROGRESS);
    }

    @Test
    void layout_Create_WithOptions() {
        final AddressLayoutConfiguration configuration = AddressLayoutConfiguration
                .newBuilder(Setup.VALID_TOKEN_ADDRESS)
                .setTransactionId(UUID.randomUUID().toString())
                .build();
        final AddressLayoutClient client = ExperianDataValidation.getAddressLayoutClient(configuration);
        //todo add logic
    }

    @Test
    void layout_Delete_DoesNotExist_Throws() {

        final AddressLayoutConfiguration configuration = AddressLayoutConfiguration
                .newBuilder(Setup.VALID_TOKEN_ADDRESS)
                .build();
        final AddressLayoutClient client = ExperianDataValidation.getAddressLayoutClient(configuration);

        final String layoutName = "ThisLayoutDoesntExist";
        assertThrows(NotFoundException.class, () -> client.deleteLayout(layoutName));

    }

    @Test
    void Format_WithCustomLayout_WithComponents() {

        final Configuration configuration = Configuration
                .newBuilder(Setup.VALID_TOKEN_ADDRESS)
                .setTransactionId(UUID.randomUUID().toString())
                .useDataset(Dataset.GB_ADDRESS)
                .setFormatLayoutName(Setup.EXISTING_TEST_LAYOUT)
                .includeComponents()
                .build();

        final Client client = ExperianDataValidation.getAddressClient(configuration);
        final com.experian.dvs.client.address.search.Result resultAutoComplete = client.search(SearchType.AUTOCOMPLETE, "56 Queens R");
        final Result formatResult = client.format(resultAutoComplete.getSuggestions().get(0).getGlobalAddressKey());
        assertThat(formatResult.getConfidence()).isEqualTo(Confidence.VERIFIED_MATCH);
        //assertThat(formatResult.getGlobalAddressKey()).isEqualTo(resultAutoComplete.getSuggestions().get(0).getGlobalAddressKey());
        assertThat(formatResult.getAddressFormatted()).isNotNull();
        assertThat(formatResult.getAddressFormatted().getLayoutName()).isEqualTo(Setup.EXISTING_TEST_LAYOUT);
        assertThat(formatResult.getAddressFormatted().hasEnoughLines()).isTrue();
        assertThat(formatResult.getAddressFormatted().hasTruncatedLines()).isFalse();
        assertThat(formatResult.getAddressFormatted().getAddress()).hasSize(4);
        assertThat(formatResult.getAddressFormatted().getAddress().get("addr_line_1")).isNotEmpty();
        assertThat(formatResult.getAddressFormatted().getAddress().get("addr_line_2")).isNotEmpty();
        assertThat(formatResult.getAddressFormatted().getAddress().get("post_code")).isNotEmpty();
        assertThat(formatResult.getAddressFormatted().getAddress().get("country_name")).isNotEmpty();
    }

    @Test
    public void suggestionsFormatWithCustomLayout() {
        final Configuration configuration = Configuration
                .newBuilder(Setup.VALID_TOKEN_ADDRESS)
                .setTransactionId(UUID.randomUUID().toString())
                .useDataset(Dataset.AU_ADDRESS)
                .setFormatLayoutName(Setup.EXISTING_TEST_LAYOUT)
                .build();
        Client client = ExperianDataValidation.getAddressClient(configuration);

        com.experian.dvs.client.address.suggestions.format.Result searchResult = client.suggestionsFormat("onslow st, perth");

        assertThat(searchResult.getSuggestions()).isNotEmpty();

        // When using a custom layout the address objects should all be empty
        assertThat(searchResult.getSuggestions()).allMatch(suggestion -> suggestion.getAddress() == null);
        assertThat(searchResult.getSuggestions()).allMatch(suggestion -> suggestion.getAddressFormatted() != null);
    }

    private static GetLayoutResult getLayout(final String layoutName) {
        final AddressLayoutConfiguration configuration = AddressLayoutConfiguration.newBuilder(Setup.VALID_TOKEN_ADDRESS).build();
        final AddressLayoutClient client = ExperianDataValidation.getAddressLayoutClient(configuration);
        return client.getLayout(layoutName);

    }

    private static void createLayout1() {

        final AddressLayoutConfiguration configuration = AddressLayoutConfiguration.newBuilder(Setup.VALID_TOKEN_ADDRESS).build();
        final AddressLayoutClient client = ExperianDataValidation.getAddressLayoutClient(configuration);

        final LayoutLineVariable line1 = new LayoutLineVariable("addr_line_1");
        final LayoutLineVariable line2 = new LayoutLineVariable("addr_line_2");
        final LayoutLineFixed line3 = new LayoutLineFixed("post_code", List.of(Aus.POSTAL_CODE, Gbr.POSTCODE));
        final LayoutLineFixed line4 = new LayoutLineFixed("country_name", List.of(Aus.COUNTRY_NAME, Gbr.COUNTRY));

        final String layoutName = Setup.EXISTING_TEST_LAYOUT;
        final CreateLayoutResult createLayoutResult = client.createLayout(layoutName,
                List.of(line1, line2),
                List.of(line3, line4),
                Dataset.AU_ADDRESS, Dataset.GB_ADDRESS);

        assertThat(createLayoutResult.getError()).isNotPresent();
    }

    private String getUniqueLayoutName() {
        return Setup.TEST_LAYOUT_PREFIX + UUID.randomUUID().toString();
    }

    private static void deleteTestLayouts() {

        //Get all layouts starting with the predefined prefix
        final AddressLayoutConfiguration configuration = AddressLayoutConfiguration.newBuilder(Setup.VALID_TOKEN_ADDRESS).build();
        final AddressLayoutClient client = ExperianDataValidation.getAddressLayoutClient(configuration);
        final GetLayoutListResult layoutsResult = client.getLayouts(Optional.empty(), List.of(), Optional.of(Setup.TEST_LAYOUT_PREFIX));

        //Delete them
        for (var layout : layoutsResult.getLayouts()) {
            if (!layout.getName().equals(Setup.EXISTING_TEST_LAYOUT) && layout.getStatus() == Status.COMPLETED) {
                try {
                    client.deleteLayout(layout.getName());
                } catch (EDVSException e) {
                    System.out.println("Failed to delete layout " + layout.getName());
                }
            }
        }
    }
}